mojo/public/tools/mojom/mojom/parse/brave_ast_patcher.py

```diff
@@ -0,0 +1,150 @@
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+"""Patches mojo modules using *.mojom files from 'chromium_src' dir."""
+
+import codecs
+import os.path
+
+from mojom.parse import ast
+from mojom.parse import conditional_features
+from mojom.parse import parser
+
+
+# Attribute constant to mark a definition as a new one.
+# If a definition with same name already exists, it's an error to add a new one.
+_DEFINITION_ADD = 1
+# Attribute constant to mark a definition as the one to extend.
+# If a definition with same name and type doesn't exist, it's an error to extend one.
+_DEFINITION_EXTEND = 2
+
+
+# Predicate for an import filename.
+def _AstImportPred(brave_import, ast_import):
+  return brave_import.import_filename == ast_import.import_filename
+
+
+# Predicate for a definition name (doesn't check type).
+def _AstDefinitionNamePred(brave_definition, ast_definition):
+  return brave_definition.mojom_name == ast_definition.mojom_name
+
+
+# Predicate for definition name and type.
+def _AstDefinitionPred(brave_definition, ast_definition):
+  return _AstDefinitionNamePred(brave_definition, ast_definition) and \
+            isinstance(brave_definition, type(ast_definition))
+
+
+# Returns required action from a definition attributes list.
+def _GetBraveDefinitionAction(brave_definition):
+  if brave_definition.attribute_list:
+    for attribute in brave_definition.attribute_list:
+      if attribute.key == 'BraveAdd':
+        return _DEFINITION_ADD
+      elif attribute.key == 'BraveExtend':
+        return _DEFINITION_EXTEND
+  raise ValueError("Definition should have [BraveAdd] or [BraveExtend] attribute: %s" %
+                   brave_definition.mojom_name)
+
+
+# Validates that a definition with same name doesn't exist in the list.
+def _CheckDefinitionDoesntExist(brave_definition, ast_items):
+  if any(_AstDefinitionNamePred(brave_definition, item) for item in ast_items):
+    raise ValueError("Definition already exists: %s" % brave_definition.mojom_name)
+
+
+# Finds a definition by name and type.
+def _FindMatchingDefinition(brave_definition, ast_definitions):
+  for ast_definition in ast_definitions:
+    if _AstDefinitionPred(brave_definition, ast_definition):
+      return ast_definition
+
+
+# Adds new items to an existing mojom definition.
+def _ExtendAstDefinition(brave_definition, ast_definition):
+  if isinstance(brave_definition, ast.Enum):
+    for item in brave_definition.enum_value_list:
+      _CheckDefinitionDoesntExist(item, ast_definition.enum_value_list)
+      ast_definition.enum_value_list.Append(item)
+  elif isinstance(brave_definition, ast.Interface) or \
+       isinstance(brave_definition, ast.Struct) or \
+       isinstance(brave_definition, ast.Union):
+    for item in brave_definition.body:
+      _CheckDefinitionDoesntExist(item, ast_definition.body)
+      ast_definition.body.Append(item)
```

Comment:
currently there is no support for nested structs, enums, constants (i.e. you can declare `struct`, `enum` inside `interface`). 
all our current changes doesn't require that, but this should be supported.

Reply:
@goodov can you comment here/create a ticket and link to it?

Reply:
I already implemented nested types support in this PR, so there is no need for ticket. Tests show all possible applications of the patcher right now, including nested types.